fathom/src/surface/elaboration.rs

```diff
@@ -1955,217 +1947,60 @@ impl<'interner, 'arena> Context<'interner, 'arena> {
-        // all the cases were covered.
-        let default_expr = match match_info.scrutinee.r#type.match_prim_spine() {
-            // No need for a default case if all the values were covered
-            Some((Prim::BoolType, [])) if branches.len() >= 2 => None,
-            _ => Some(self.elab_match_absurd(is_reachable, match_info)),
+        pattern: &Pattern<'_, ByteRange>,
+        r#type: Option<&Term<'_, ByteRange>>,
+    ) -> (CheckedPattern<'arena>, Scrutinee<'arena>) {
+        let (pattern, r#type) = self.synth_ann_pattern(pattern, r#type);
+        let expr = core::Term::LocalVar(pattern.range().into(), env::Index::last());
+        let scrut = Scrutinee {
+            range: pattern.range(),
+            expr: self.scope.to_scope(expr),
+            r#type,
         };
-
-        core::Term::ConstMatch(
-            full_span,
-            match_info.scrutinee.expr,
-            self.scope.to_scope_from_iter(branches.into_iter()),
-            default_expr.map(|expr| (None, self.scope.to_scope(expr) as &_)),
-        )
+        (pattern, scrut)
     }
 
-    /// Elaborate unreachable match cases. This is useful for that these cases
-    /// are correctly typed, even if they are never actually needed.
-    fn elab_match_unreachable<'a>(
+    fn check_param(
         &mut self,
-        match_info: &MatchInfo<'arena>,
-        equations: impl Iterator<Item = &'a (Pattern<ByteRange>, Term<'a, ByteRange>)>,
-    ) {
-        self.elab_match(match_info, false, equations);
+        pattern: &Pattern<'_, ByteRange>,
+        r#type: Option<&Term<'_, ByteRange>>,
+        expected_type: &ArcValue<'arena>,
+    ) -> (CheckedPattern<'arena>, Scrutinee<'arena>) {
+        let pattern = self.check_ann_pattern(pattern, r#type, expected_type);
+        let expr = core::Term::LocalVar(pattern.range().into(), env::Index::last());
+        let scrut = Scrutinee {
+            range: pattern.range(),
+            expr: self.scope.to_scope(expr),
+            r#type: expected_type.clone(),
+        };
+        (pattern, scrut)
     }
 
-    /// All the equations have been consumed.
-    fn elab_match_absurd(
+    fn elab_match(
         &mut self,
-        is_reachable: bool,
-        match_info: &MatchInfo<'arena>,
+        mut matrix: PatMatrix<'arena>,
+        bodies: &[Body<'arena>],
+        match_range: ByteRange,
+        scrut_range: ByteRange,
     ) -> core::Term<'arena> {
-        // Report if we can still reach this point
-        if is_reachable {
-            // TODO: this should be admitted if the scrutinee type is uninhabited
-            self.push_message(Message::NonExhaustiveMatchExpr {
-                match_expr_range: match_info.range,
-                scrutinee_expr_range: match_info.scrutinee.range,
-            });
-        }
-        core::Term::Prim(match_info.range.into(), Prim::ReportedError)
+        debug_assert_eq!(matrix.num_rows(), bodies.len());
```

Comment:
I’d be cool if we added [`contracts`](https://lib.rs/crates/contracts) to attach preconditions like this!

Reply:
In my experience, `contracts` can clutter backtraces and may cause compile errors in functions with lifetime annotations.